cpp/src/community/leiden_impl.cuh

```diff
@@ -466,34 +467,48 @@ std::pair<std::unique_ptr<Dendrogram<vertex_t>>, weight_t> leiden(
           std::make_optional<edge_property_view_t<edge_t, weight_t const*>>(
             (*coarsen_graph_edge_weight).view());
 
-        // cluster_assignment contains leiden cluster ids of aggregated nodes
-        // After call to relabel, cluster_assignment will louvain cluster ids
-        // of the aggregated nodes
-        relabel<vertex_t, multi_gpu>(
-          handle,
-          std::make_tuple(static_cast<vertex_t const*>(leiden_to_louvain_map.first.begin()),
-                          static_cast<vertex_t const*>(leiden_to_louvain_map.second.begin())),
-          leiden_to_louvain_map.first.size(),
-          (*cluster_assignment).data(),
-          (*cluster_assignment).size(),
-          false);
-        // louvain assignment of aggregated graph which is necessary to flatten dendrogram
-        dendrogram->add_level(current_graph_view.local_vertex_partition_range_first(),
-                              current_graph_view.local_vertex_partition_range_size(),
-                              handle.get_stream());
+        // FIXME: reconsider what's put into dendrogram->current_level_begin()
+        // at what point in the code. I'm just going to overwrite it here,
+        // so perhaps it should be in different structures until now
 
+        // New approach, mimic Louvain, we'll store the Leiden results in the dendrogram
         raft::copy(dendrogram->current_level_begin(),
-                   (*cluster_assignment).begin(),
-                   (*cluster_assignment).size(),
+                   refined_leiden_partition.data(),
+                   refined_leiden_partition.size(),
                    handle.get_stream());
 
         louvain_of_refined_graph.resize(current_graph_view.local_vertex_partition_range_size(),
                                         handle.get_stream());
 
+        // Temporarily use louvain_of_refined_graph to be a numeric sequence to renumber the
+        // dendrogram.
+        detail::sequence_fill(handle.get_stream(),
+                              louvain_of_refined_graph.data(),
+                              louvain_of_refined_graph.size(),
+                              current_graph_view.local_vertex_partition_range_first());
+
+        relabel<vertex_t, multi_gpu>(
+          handle,
+          std::make_tuple(static_cast<vertex_t const*>((*numbering_map).begin()),
+                          static_cast<vertex_t const*>(louvain_of_refined_graph.begin())),
+          current_graph_view.local_vertex_partition_range_size(),
```

Comment:
` (*numbering_map).size() `<= `current_graph_view.local_vertex_partition_range_size()`
 I think we need to use ` (*numbering_map).size() ` here

Reply:
When can `(*numbering_map).size() < current_graph_view.local_vertex_partition_range_size()`???

Don't miss https://github.com/rapidsai/cugraph/pull/4347/files#diff-609b314aeaec6ca07d296e41bdf9f67407b5ffe539ce568506290925e9b169abR464

I think `(*numbering_map).size() == current_graph_view.local_vertex_partition_range_size()`

Reply:
You are are right, they are equal.